[copy-propagation] Extract out CanonicalOSSALifetime::computeCanonicalLiveness/its state into a helper struct CanonicalOSSALifetimeInfo.

gottesmm opened the pull request:
The intention is that with this refactoring, I can reuse this analysis in a new
_move operator diagnostic for copyable types.

atrick reviewed (approved):
I don't know exactly what this API should be in the long term. I want to rewrite it to reuse PrunedLivenessBoundary and simplify the rewrite logic. But this seems like a fine refactoring to make forward progress.

There are a couple of stray changes that don't make sense to me

gottesmm committed:
[moveOnly] Add checker that validates that if a let/param copyable value has _move applied to it, the let doesn't have any later uses.

This is just an initial prototype for people to play with. It is as always
behind the -enable-experimental-move-only flag.

NOTE: In this PR I implemented this only for 'local let' like things (local
lets/params). I did not implement in this PR support for local var and haven't
done anything with class ivars or globals.

rdar://83957028

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -269,34 +361,9 @@ class CanonicalizeOSSALifetime {
   GraphNodeWorklist<SILValue, 8> defUseWorklist;
 
-  /// Visited set general CFG traversal that prevents revisiting blocks.
-  GraphNodeWorklist<SILBasicBlock *, 8> blockWorklist;
```
@atrick, @eeckstein I tried to not extract out this blockWorklist into CanonicalOSSALifetimeInfo like I did in this patch and instead use a BasicBlockWorklist for the relevant computations. I kept on running into weird behavior and I couldn't figure out what it was. So by moving this around a little bit, I was able to just keep the code as it was. I will mark the function below where I wanted to use BlockWorklist. I didn't have time to follow up with what is up with it.

gottesmm commented on lib/SILOptimizer/Utils/CanonicalOSSALifetime.cpp:
```diff
@@ -295,13 +296,16 @@ endsAccessOverlappingPrunedBoundary(SILInstruction *inst) {
+
   while (changed) {
     changed = false;
     blockWorklist.initializeRange(consumingBlocks);
```
@eeckstein @atrick This is the place where I ran into weird memory behavior like weirdness behavior when having a BasicBlockWorklist.

The changes I made were:

1. I put in the check above that we hit the loop with at least one consuming block so we always have at least one block that we can use to initialize BasicBlockWorklist. (I subsequently kept that check if empty since I think it is a nice invariant and with time I think we will want to use BasicBlockWorklist in some form here.
2. I changed initializeRange/insert to pushIfNotVisited.

Shouldn't that just work?

gottesmm commented on include/swift/SIL/BasicBlockDatastructures.h:
```diff
@@ -101,6 +101,13 @@ class BasicBlockWorklist {
   }
 
+  /// Initialize the worklist with \p initialBlock. Initial blocks must not be
+  /// empty!
```
I redid this API.

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -373,19 +431,31 @@ class CanonicalizeOSSALifetime {
+  decltype(livenessInfo.debugValues) &getDebugValues() {
+    return livenessInfo.debugValues;
+  }
+  decltype(livenessInfo.liveness) &getLiveness() {
```
Good point! In an update, I eliminated all of the decltype here. We should optimize for reading!

gottesmm commented:
@swift-ci test

gottesmm commented on include/swift/SILOptimizer/PassManager/Passes.def:
```diff
@@ -418,6 +418,9 @@ PASS(AssemblyVisionRemarkGenerator, "assembly-vision-remark-generator",
      "Emit assembly vision remarks that provide source level guidance of where runtime calls ended up")
 PASS(MoveOnlyChecker, "sil-move-only-checker",
      "Pass that enforces move only invariants on raw SIL")
+PASS(CopyableValueMoveOperatorChecker, "sil-copyable-value-move-operator-checker",
```
I included this by mistake. I chopped this PR off of a much larger out of tree working copy that I am "slicing" PRs off of. Sorry!

gottesmm commented:
I did this a different way.
